src/hooks/usePrevious.ts

```diff
@@ -1,14 +1,18 @@
-import { useEffect, useRef } from 'react'
+import { useEffect, useState } from 'react'
 
 /** Retains & returns the previous value passed in each time the hook is invoked. */
-function usePrevious<T>(value: T): T | undefined {
-  const ref = useRef<T>()
+function usePrevious<T>(value: T): T {
+  const [current, setCurrent] = useState(() => value)
+  const [previous, setPrevious] = useState(() => value)
 
   useEffect(() => {
```

Comment:
@raineorshine One thing that I didn't expected to see: this hook gets unmounted (a returned callback is called) at various points during the render cycle. I believe that caused it to run again even if `value` hadn't changed.

Reply:
It doesn't seem to be the entire component, just the effect inside the hook. Maybe I was misunderstanding something.

Reply:
Yeah, I'm not sure.